### ISSUE TITLE
ALEpDist p-values problem

### ISSUE THREAD

**Issue by snvv** [tripartio/ale#14]
#### Title
ALEpDist returns p-values of exactly 0 or 1 when using `gpboost` with custom prediction function

#### Description
I am trying to calculate feature significance (p-values) for an additive model combining gradient boosting (`gpboost`) with random effects via the `ale::ALEpDist()` function. However, regardless of the configuration, the resulting p-values are always either exactly `0` or `1`. This behavior persists even after multiple attempts to adjust settings and rework the custom prediction logic.

I suspect this may be related to how predictions are handled between `gpboost` and `ALEpDist`, particularly around group data or random effect handling, but I have not been able to identify the root cause.

This makes it impossible to interpret variable importance using statistical significance.

---

#### Minimal Reproducible Example

```r
# Load required packages
library(gpboost)
library(ale)
library(dplyr)

# Define %||% operator if not using rlang
`%||%` <- function(x, y) if (is.null(x)) y else x

# Set seed for reproducibility
set.seed(123)

# Use var_cars dataset from ale package
data(var_cars, package = "ale")

# Prepare data
data <- var_cars %>% 
  select(mpg, hp, wt, am, cyl)

# Calculate group sizes and labels for random effects
group_sizes <- as.integer(table(data$cyl))
group_data <- data$cyl

# Prepare gpb.Dataset
ds <- gpb.Dataset(
  data = as.matrix(data[, c("hp", "wt", "am")]),
  label = data$mpg
)

# Define GP model for random effects
gp_model_re <- GPModel(
  group_data = group_data,
  likelihood = "gaussian"
)

# Train gpboost model with fixed and random effects
gp_model <- gpb.train(
  data = ds,
  gp_model = gp_model_re,
  params = list(
    objective = "regression",
    learning_rate = 0.1,
    max_depth = 3,
    num_leaves = 7,
    num_iterations = 100,
    min_data_in_leaf = 5,
    verbose = 0
  )
)

# Custom predict function
custom_predict <- function(object, newdata, type = "response") {
  base_cols <- c("hp", "wt", "am")
  if (!all(base_cols %in% colnames(newdata))) {
    stop("Missing required columns in newdata: ", paste(base_cols[!base_cols %in% colnames(newdata)], collapse = ", "))
  }
  
  group_data_pred <- if ("cyl" %in% colnames(newdata)) newdata$cyl else rep(data$cyl[1], nrow(newdata))
  X_new <- as.matrix(newdata[, base_cols])
  
  expected_features <- object$params$num_features %||% 3
  if (expected_features == 4) {
    X_new <- cbind(X_new, random_variable = rep(0, nrow(newdata)))
  }

  pred <- predict(object, data = X_new, group_data_pred = group_data_pred, predict_disable_shape_check = TRUE)$response_mean
  return(as.numeric(pred))
}

# Test custom prediction
test_preds <- custom_predict(gp_model, data)

# Random model call string for exact p-values
random_model_call_string <- '{
  cat("ALEpDist: nrow(rand_data) =", nrow(rand_data), "\\n")
  cat("ALEpDist: unique(rand_data$cyl) =", unique(rand_data$cyl), "\\n")
  model <- gpboost::gpb.train(
    data = gpb.Dataset(
      data = as.matrix(rand_data[, c("hp", "wt", "am", "random_variable")]),
      label = rand_data$mpg
    ),
    gp_model = gpboost::GPModel(
      group_data = rand_data$cyl,
      likelihood = "gaussian"
    ),
    params = list(
      objective = "regression",
      learning_rate = 0.1,
      max_depth = 3,
      num_leaves = 7,
      num_iterations = 100,
      min_data_in_leaf = 5,
      verbose = 0
    )
  )
  model
}'

# Compute ALE without p-values
ale_result <- ALE(
  model = gp_model,
  x_cols = c("hp", "wt", "am"),
  data = data,
  y_col = "mpg",
  pred_fun = custom_predict,
  pred_type = "response",
  boot_it = 0,
  p_values = NULL,
  silent = FALSE,
  parallel = 0
)

# Compute ALEpDist for exact p-values
ale_pdist <- ALEpDist(
  model = gp_model,
  data = data,
  y_col = "mpg",
  x_cols = c("hp", "wt", "am"),
  pred_fun = custom_predict,
  pred_type = "response",
  rand_it = 100,
  random_model_call_string = random_model_call_string,
  random_model_call_string_vars = c("hp", "wt", "am"),
  parallel = "all but one",
  silent = FALSE
)

# Inspect results
print(ale_pdist@params)
print(as.data.frame(ale_pdist@rand_stats$mpg))

# Attach p-values to ALE
ale_with_pvalues <- ALE(
  model = gp_model,
  x_cols = c("hp", "wt", "am"),
  data = data,
  y_col = "mpg",
  pred_fun = custom_predict,
  pred_type = "response",
  p_values = ale_pdist,
  boot_it = 0,
  silent = FALSE,
  parallel = 0
)

# Plot final ALE with p-values
plot(ale_with_pvalues)
```

#### Expected Behavior

- `ALEpDist` should return meaningful p-values (between 0 and 1) indicating the statistical significance of each feature.

---

#### Actual Behavior

- All returned p-values are exactly `0` or `1`.
- This occurs both with the `gpboost` model directly and with surrogate models like linear regression.
- Debugging suggests that permutations may not be affecting the output enough to produce variation in effect distributions.

---

#### Environment

- R version:  4.5.0-2 
- Operating System:  debian sid 

Any help or suggestions would be **greatly appreciated**!

Thank you,  
SN

**Comment by tripartio** [comment]
I'm not sure if I'll have the time to try to debug this right away since I'll be traveling or on vacation for most of next week. (May is a month of holidays here in France). 

**But right away, please try to run this analysis on a real dataset** like `ggplot2::diamonds` (regression) or `ale::census` (classification). `ale::vars_cars` is a tiny toy dataset, and almost any machine learning model run on such a tiny dataset will be a mess. That alone might fix your problem. If not, please replace your MRE with one of those datasets of real data.

**Comment by tripartio** [comment]
Also, here's a speed hint: try to make your custom prediction function as efficient as possible. It is the single slowest part of the ALE algorithm, so making it more efficient will noticeably speed everything up. In particular, skip error checking and any other step that is not absolutely necessary. (The ALE constructor has built-in checks for errors in the custom predict function, so you don't need to slow it down with additional error checking.)

**Comment by snvv** [comment]
I'm pretty sure that the examples you told me to try will run without a problem. The problem is with the specific program and data structure. At this stage I am not interested in speed but in solving the problem. I will wait to see the problem when you get back. 
Have a great time on your vacation!!!

**Comment by tripartio** [comment]
> I'm pretty sure that the examples you told me to try will run without a problem. The problem is with the specific program and data structure. At this stage I am not interested in speed but in solving the problem. I will wait to see the problem when you get back. Have a great time on your vacation!!!

I'm sorry, but if the problem is that you run into incomprehensible results when you inappropriately run a machine learning model on a tiny dataset that is way too small to be valid, then I won't even bother to try to debug it. If that's the case, then the problem has nothing to do with ALE or the `ale` package; it is a modelling design error. If you have problems with a legitimate model, I'll be happy to try to help you, but not with a blatantly invalid model that violates assumptions like minimum appropriate sample size. I hope that makes sense.

Why not try it out on your real application data and make sure it works correctly with that? That's what I try to help people with.
